Make autocast cache and buffer stealing aware of cudagraph static output tensors

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99101
* __->__ #99368

In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

eellison committed:
cpp changes for cached cuda graph tensors

[ghstack-poisoned]

ngimel commented on aten/src/ATen/autocast_mode.cpp:
```diff
@@ -165,6 +177,35 @@ void set_autocast_cache_enabled(bool enabled) {
+}
+
+void set_cudagraph_cached_tensors_enabled(bool enabled) {
+  cudagraph_cached_tensorimpl_set_enabled = enabled;
```
without locks and without static?

eellison replied:
just as with autocast enabled, the global boolean will be set true before the map is written or read from, so you dont need a lock.

eellison replied:
I can add static... I saw that `cached_casts` did not use that, but maybe that's an omission ?

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented:
This PR seems a bit finely balanced (adjusting refcounts so the heuristics fire the same way both times.) Weren't we just straight up disabling autocast cache at some point? Why does that recipe not apply here?

eellison commented:
@ezyang I need to fix the problem of gradient inplacing differing between execution and record due to `can_accumulate_inplace` not firing because of the extra reference. The fix for that also fixes the cached cast cast problem. I had proposed the other fix for amp before I discovered the gradient inplacing issue. We do a similar ref count adjusting in `accumulateGrad`, for what it's worth.

ezyang commented:
This still doesn't make sense to me. We turn off multithreading on tracing. Why not turn off the in place buffer update too? Remember that we are going to functionalize it anyway!

eellison commented:
@ezyang turning off global buffer in placing is a perf and memory hit

eellison commented:
@ezyang we don't have any control of these code paths being hit outside of tracing. We already turn off caching when we trace with amp but that didn't prevent this bug being hit. @ngimel also told me that turning off buffer inplacing would not be good for perf and memory. I don't think it's great that using cudagraphs in a small part of your model would lead to a global change of behavior elsewhere.

ezyang commented:
I talked about it with Elias in person. The key distinction is that the buffer stealing is happening *outside* of cuda graph tracing, but it affects subsequent tracing because doing an inplace versus out of place is the distinction between the tensor showing up later as from cuda graph tensor or not.

Elias, I rereviewed the PR description, and it doesn't spell this out clearly enough, I think. A worked example would go a long way here.

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -5,6 +5,7 @@
 #include <ATen/TensorOperators.h>
 #include <ATen/TensorSubclassLikeUtils.h>
 #include <ATen/native/SparseTensorUtils.h>
+#include <torch/csrc/autograd/functions/accumulate_grad.h>
```
This should not depend on accumulate grad.
If you need a shared util, you can create a utils file.

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -77,7 +78,8 @@ bool can_accumulate_inplace(const Variable& v) {
 
       // and we hold the last reference
-      v.use_count() == 1 && v.has_storage() && v.storage().use_count() == 1);
+      torch::autograd::adjusted_use_count(v) == 1 && v.has_storage() &&
```
Could you explain why this is ok to do?
This will make it so that the autograd engine will inplace modify the Tensor in question. So I'm pretty sure the "caching" of its convertion across dtype will be invalid!

eellison replied:
We're disabling the caching of dtype. https://github.com/pytorch/pytorch/pull/99368/files#diff-b99d3aff6a9039f8abada80fa8c2647a85d8248fa22b64310a0cc9ca620dad62R216-R220

albanD commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -113,7 +118,7 @@ struct TORCH_API AccumulateGrad : public Node {
           !new_grad.is_sparse_csr() &&
           !(variable.is_sparse_csr() && new_grad.layout() == at::kStrided) &&
-          new_grad.use_count() <= num_expected_refs &&
+          torch::autograd::adjusted_use_count(new_grad) <= num_expected_refs &&
```
Why is this ok?
Once this get returned to the user, they are free to do whatever they want with it. In particular any inplace update! So you would need to properly invalidate any caching related to this Tensor (which you can't do).

eellison replied:
If a user does caching of a TensorImpl outside of our system, that is a general limitation of cudagraphs, and not related to this pr

albanD replied:
But this check is to choose if the TensorImpl is used as-is in the `.grad` field. So it is expected that the user will use this .grad outside of your system!

eellison replied:
For both warmup and recording, we are hitting this path, and for both of those we have the same limitation that we outputs from one generation cannot be present for the next generation. So this is already the current state. We could try ban accumulation for cudagraphs outputs but it's not clear that the new grad wouldn't be a cudagraph output either. I would prefer to do that in a separate PR.

ngimel replied:
Would it be cleaner if you instead adjusted num_expected_refs at the callsite?

albanD reviewed (changes requested):
Doesn't look sound to me.

If it turns out to be and we want to do this. I don't think the new dependencies are ok and they would need to be cleaned up.

eellison commented on aten/src/ATen/autocast_mode.cpp:
```diff
@@ -73,9 +73,21 @@ namespace {
+// we hash on the TensorImpl* and keep the pointer alive with a weakref value.
+bool cudagraph_cached_tensorimpls_enabled = false;
+ska::flat_hash_map<TensorImpl*, weakref_type> cudagraph_cached_tensorimpls;
+std::mutex cudagraph_cached_tensorimpl_mutex;
```
There's no need to increase contention around two unrelated mutexes

eellison replied:
in either case, I'm moving to separate file/util..

eellison commented on aten/src/ATen/autocast_mode.cpp:
```diff
@@ -73,9 +73,21 @@ namespace {
 std::mutex cached_casts_mutex;
 
+
+// Some CUDA Graph Implementations will persist a static tensor output whose TensorImpl
```
I can remove the cuda graph specific references.

eellison commented on test/test_autocast.py:
```diff
@@ -125,6 +125,46 @@ def test_autocast_torch_need_autocast_promote(self):
+
+    def __enter__(self):
+        self.old_clear_cache = torch.clear_autocast_cache
+        torch.clear_autocast_cache = lambda: None
```
This is existing test code which I'm refactoring.

albanD replied:
Sure, I was just curious why this is used here?

eellison replied:
I'm testing that we are not caching the weight, and reusing the infra across the existing test and new test.. not sure I understand the question

eellison commented:
> Elias, I rereviewed the PR description, and it doesn't spell this out clearly enough, I think. A worked example would go a long way here.

None of the original Prs for buffer accumulation have tests, so it's hard to replicate

albanD commented:
You have quite a few tests in `test/test_autograd.py` for this. In particular `test_accumulate*` and I'm sure there are others in this file and others depending on the specific topic (for the layout spec, sparse, subclass etc)

albanD commented:
This whole thing would not need to exist if we could use https://github.com/pytorch/pytorch/blob/d8d479c854c622a6ef21190b7e62c9a76f0ea2a7/torch/utils/weak.py#L85 for the map within autocast...

eellison commented:
@albanD we are already doing a weak cache in autocast. how does that help?

albanD commented:
You don't hold a strong reference so you don't need special logic in other subsystems to ignore that strong reference.

eellison commented:
We need the strong reference to avoid reconstructing tensors on the hot path in re running cudagraph replay. That's where the strong reference comes from.

albanD commented:
I'm really confused at this point as everything has cudagraph in its name here...
If if you're saying the cudagraph_tree.py data structure is what holds onto strong reference, why are we checking that through some autocast data structure?

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented:
> If if you're saying the cudagraph_tree.py data structure is what holds onto strong reference, why are we checking that through some autocast data structure?

I've moved it to a separate utils file as you suggested. The reasons why we need to check if there is the additional strong reference in cudagraph_trees are explained in the pr description. happy to vc if that would be more helpful.

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented:
cc @malfet @albanD would you advise on how I should fix mobile builds here ? I would have thought the ifdef mobile would be sufficient. it's always a pain adding new files 😦

ezyang commented on aten/src/ATen/cached_tensor_utils.h:
```diff
@@ -0,0 +1,33 @@
+#include <ATen/ATen.h>
+
+namespace at {
+namespace caching {
```
This file needs more docs

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented on aten/src/ATen/CachedTensorUtils.h:
```diff
@@ -0,0 +1,26 @@
+// flag will be false and we will avoid the hash lookup.
+
+TORCH_API bool is_cached_tensor(const at::Tensor& t);
+TORCH_API void add_cached_tensor(const at::Tensor& t);
```
What would happen if you add_cached_tensor a tensor twice?

eellison replied:
We're hashing on the TensorImpl, would be the same as inserting once

ezyang replied:
But then if you remove it, I clear all of them? That seems troublesome? Wouldn't it make more sense for any tensor to only be addable once?

eellison replied:
It's a set and it has the same semantics as a set. I added a test to verify that it works as intended.

ngimel commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -130,9 +131,9 @@ struct TORCH_API AccumulateGrad : public Node {
-          new_grad._indices().use_count() <= 1 &&
-          new_grad._values().use_count() <= 1 &&
-          new_grad.use_count() <= num_expected_refs) {
+          at::caching::adjusted_use_count(new_grad._indices()) <= 1 &&
```
can sparse tensor indices and values ever be added to cache? I though cudagraph trees would require special handling for sparse tensors for this to happen, and right now trees just won't handle sparse tensors correctly?

eellison replied:
Yea, good point!

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

albanD commented on aten/src/ATen/CachedTensorUtils.cpp:
```diff
@@ -0,0 +1,49 @@
+
+
+namespace at {
+namespace caching {
```
This is not doing any caching so it is very weird to use such a namespace for it?

eellison replied:
what would you suggest ?  I had something more cudagraph specific originally.

albanD commented on aten/src/ATen/CachedTensorUtils.cpp:
```diff
@@ -0,0 +1,49 @@
+  cached_tensorimpls_enabled = enabled;
+}
+
+size_t adjusted_use_count(const at::Tensor& t) {
```
Why doesn't this live directly on the Tensor? In which case do you expect that we don't want to use this value?

albanD reviewed (changes requested):
I still feel like this change is quite more complex than it needs to be and it doesn't make a lot of sense when removing the cudagraph lens.

> . The reasons why we need to check if there is the additional strong reference in cudagraph_trees are explained in the pr description.

First, Is the strong reference always a requirement? Because the issue you're trying to avoid here is about having 2+ strong references to these Tensors. In such a case, a strong reference in cudagraph_trees is not needed right?

Second, is caching the exact same TensorImpl a requirement for your cudagraph work? Or would re-creating a new TensorImpl (that wraps the same storage) for each call be an ok overhead for you? That might save you a lot of trouble.

Finally if we are ready to take a namespace like `at::caching` for this, then we might as well make this a method on Tensor (we have space in the bitfield for this) and remove the need for the hash_map and new file altogether. You can then either update the original use_count to always return the biased value or add a new method to get it.

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented:
We discussed offline, but to restate here:

> First, Is the strong reference always a requirement? Because the issue you're trying to avoid here is about having 2+ strong references to these Tensors. In such a case, a strong reference in cudagraph_trees is not needed right?

The strong reference in cudagraph_trees is necessary to avoid overhead. For some small, overhead bound networks this is a significant portion of runtime. The way dynamo (and aot autograd) work is by separating forward and backward, so even if you are cudagraphing the whole thing you still need to materialize tensor outputs.

> Second, is caching the exact same TensorImpl a requirement for your cudagraph work? Or would re-creating a new TensorImpl (that wraps the same storage) for each call be an ok overhead for you? That might save you a lot of trouble.

I don't think this works, as [can_accumulate_inplace](https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/input_buffer.cpp#L71-L80) checks Storage count. I think it is likely is a bug that [accumulate_grad](https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/functions/accumulate_grad.h#L116) doesn't do the same check.

Do you have a suggestion as to what namespace should be ? When I originally had cudagraph related names, I was told to do something more generic.

albanD reviewed (approved):
From more offline discussion, sounds like we have to do this at this point.

eellison committed:
Update on "Make autocast cache and buffer stealing aware of cudagraph static output tensors"


In this stack of PRs we adding caching to output tensors for cudagraph trees after we've done initial recording. On initial recording we do not cache tensor outputs because this prevents memory from being reclaimed. On subsequent exeuctions we do cache them to avoid overhead. However, because there is an extra reference around, this caused divergent recording & execution behavior in both autocast caching and autograd gradient stealing. Divergent recording & execution would keep on re-recording and eventually stabilize, but it's not what you want to see happen. 

This pr makes the autocast cache and buffer stealing aware of the cudagraph static output tensors. 

I will add this to the other cudagraph impl in another pr.

Not sure if this should be in autograd or in autocast since it affects both.. Or somewhere else

cc mcarilli ptrblck leslie-fang-intel jgong5 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
